src/yarn/typescript-workspace.ts

```diff
@@ -290,9 +290,26 @@ export class TypeScriptWorkspace extends typescript.TypeScriptProject implements
       isPrivatePackage: this.isPrivatePackage,
       // Empty string leads to default behavior of ^, otherwise we specify an exact version of 0.0.0
       // which will be replaced come release time.
-      dependencyRange: refOpts?.exactVersion ? '0.0.0' : '',
+      dependencyRange: refOpts?.exactVersion ? this.actualPackageVersion : '',
     };
   }
+
+  /**
+   * Return the actual declared version number from this package's package.json
+   *
+   * If that file doesn't exist, it's assumed to be `0.0.0`.
+   */
+  public get actualPackageVersion(): string {
+    try {
+      const pkgJson = JSON.parse(fs.readFileSync(join(this.outdir, 'package.json'), 'utf-8'));
+      return pkgJson.version;
+    } catch (e: any) {
+      if (e.code === 'ENOENT') {
```

Comment:
Why not check for file existence?

Reply:
In general, if you want to do an operation it's better to do the operation and check for failure, than checking if you're pretty sure that it might work if you tried it, then try it and not check for the failure.

See also [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use)